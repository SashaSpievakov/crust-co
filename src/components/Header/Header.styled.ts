import styled, { css } from 'styled-components';

// import { Container } from '../../styles/Base.styled';

interface WrapperProps {
  isCart: boolean;
}

interface LiProps {
  mode: string;
}

const HeaderWrapper = styled.header<WrapperProps>`
  position: relative;
  width: 90%;
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin: 0 auto;
  padding: 40px 0;
  border-bottom: 1px solid ${({ theme }) => theme.colors.gray};

  @media screen and (max-width: 550px) {
    padding: 30px 0;
  }

  @media screen and (max-width: 400px) {
    gap: 15px;
  }

  ${({ isCart }) =>
    !isCart &&
    css`
      @media screen and (max-width: 670px) {
        flex-direction: column;
        align-items: start;
        row-gap: 20px;
      }
    `}
`;

const Right = styled.section`
  display: flex;
  align-items: center;
  column-gap: 30px;

  @media screen and (max-width: 670px) {
    flex-direction: row-reverse;
  }
`;

const Icon = styled.i<LiProps>`
  font-size: 23px;
  color: ${({ theme }) => theme.colors.third};
  user-select: none;
  cursor: pointer;

  &:hover {
    color: ${({ theme }) => theme.colors.primary};
  }

  @media screen and (max-width: 800px) {
    font-size: 20px;
  }

  ${({ mode }) =>
    mode === 'cart' &&
    css`
      @media screen and (max-width: 450px) {
        position: absolute;
        font-size: 23px;
        top: -18px;
      }
    `}
`;

const Delimiter = styled.div`
  width: 1px;
  height: 25px;
  background-color: rgba(255, 255, 255, 0.25);
  margin-left: 14px;
  margin-right: 14px;
`;

export { HeaderWrapper, Right, Icon, Delimiter };
