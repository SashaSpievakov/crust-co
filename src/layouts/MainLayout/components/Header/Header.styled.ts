import { Button } from 'src/styles/Buttons.styled';
import styled, { css } from 'styled-components';

interface HeaderProps {
  isCart: boolean;
}

interface LiProps {
  mode: string;
}

const HeaderWrapper = styled.header<HeaderProps>`
  position: relative;
  width: 90%;
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin: 0 auto;
  padding: 40px 0;

  @media screen and (max-width: 550px) {
    padding: 30px 0;
  }

  @media screen and (max-width: 400px) {
    gap: 15px;
  }

  ${({ isCart }) =>
    !isCart &&
    css`
      @media screen and (max-width: 670px) {
        flex-direction: column;
        align-items: start;
        row-gap: 20px;
      }
    `}
`;

const Right = styled.section`
  display: flex;
  align-items: center;
  column-gap: 30px;

  @media screen and (max-width: 670px) {
    flex-direction: row-reverse;
  }
`;

const Icon = styled.i<LiProps>`
  font-size: 23px;
  color: ${({ theme }) => theme.colors.third};
  user-select: none;
  border-radius: 5px;
  cursor: pointer;

  &:hover {
    color: ${({ theme }) => theme.colors.primary};
  }

  &:focus-visible {
    outline: 4px solid ${({ theme }) => theme.colors.secondary};
    outline-offset: 5px;
  }

  ${({ mode }) =>
    mode === 'cart' &&
    css`
      @media screen and (max-width: 450px) {
        position: absolute;
        font-size: 23px;
        top: 15px;
      }
    `}
`;

const ButtonCart = styled(Button)`
  display: flex;
  align-items: center;
  line-height: 23px;
  padding: 12px 25px;

  svg {
    margin-right: 8px;
    margin-bottom: 1px;
  }

  span {
    font-weight: 600;
    font-size: 16px;
  }

  &:focus-visible {
    outline: 5px solid #d5d5d5;
  }
`;

const Delimiter = styled.div`
  width: 1px;
  height: 25px;
  background-color: rgba(255, 255, 255, 0.25);
  margin-left: 14px;
  margin-right: 14px;
`;

export { ButtonCart, Delimiter, HeaderWrapper, Icon, Right };
