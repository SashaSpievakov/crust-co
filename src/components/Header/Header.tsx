import { useEffect, useRef } from 'react';
import { shallowEqual } from 'react-redux';
import { Link, useLocation } from 'react-router-dom';
import { BsCart3, BsFillSunFill, BsFillMoonFill } from 'react-icons/bs';

import LogoSection from '../UI/LogoSection/LogoSection';
import { setTheme } from '../../store/slices/theme/reducer/themeReducer';
import { selectIsLight } from '../../store/slices/theme/selectors/selectIsLight';
import { selectCart } from '../../store/slices/cart/selectors/selectCart/selectCart';
import { Main, Wrapper, Right, Icon, Delimiter } from './Header.styled';
import { ButtonCart } from '../../styles/Buttons.styled';
import { useAppDispatch, useAppSelector } from '../../hooks/reduxHooks';

const Header = () => {
  const scrollRef = useRef<null | HTMLDivElement>(null);
  const isMounted = useRef<boolean>(false);
  const isMountedSecond = useRef<boolean>(false);
  const location = useLocation();

  const { itemsCount, totalPrice, items } = useAppSelector(
    selectCart,
    shallowEqual,
  );
  const isLight = useAppSelector(selectIsLight);
  const dispatch = useAppDispatch();

  // scroll effect
  useEffect(() => {
    if (isMountedSecond.current) {
      scrollRef.current?.scrollIntoView();
    }

    isMountedSecond.current = true;
  }, [location.pathname]);

  // local storage effects
  useEffect(() => {
    if (isMounted.current) {
      const jsonCart = JSON.stringify(items);
      const jsonTheme = JSON.stringify(isLight);

      localStorage.setItem('cart', jsonCart);
      localStorage.setItem('isLight', jsonTheme);
    }

    isMounted.current = true;
  }, [items, isLight]);

  const changeThemes = () => {
    dispatch(setTheme(!isLight));
  };

  return (
    <Main ref={scrollRef}>
      <Wrapper isCart={location.pathname === '/cart'}>
        <LogoSection />

        <Right>
          <Icon
            as={isLight ? BsFillMoonFill : BsFillSunFill}
            onClick={changeThemes}
            mode={location.pathname === '/cart' ? 'cart' : ''}
            data-testid="themeIcon"
          />
          {location.pathname !== '/cart' && (
            <ButtonCart to="/cart" as={Link} data-testid="cartLink">
              <span>{totalPrice} $</span>
              <Delimiter />
              <BsCart3 />
              <span>{itemsCount}</span>
            </ButtonCart>
          )}
        </Right>
      </Wrapper>
    </Main>
  );
};

export default Header;
