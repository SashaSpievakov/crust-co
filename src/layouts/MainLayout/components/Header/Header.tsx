import { FC, KeyboardEvent, useEffect, useRef } from 'react';
import { BsCart3, BsFillMoonFill, BsFillSunFill } from 'react-icons/bs';
import { shallowEqual } from 'react-redux';
import { Link, useLocation } from 'react-router-dom';
import { useAppDispatch, useAppSelector } from 'src/hooks';
import { selectIsLight, setTheme } from 'src/store';
import { selectCart } from 'src/store/slices/cart/selectors/selectCart/selectCart';
import { ButtonCart } from 'src/styles/Buttons.styled';

import { Delimiter, HeaderWrapper, Icon, Right } from './Header.styled';
import { LogoSection } from './LogoSection';

export const Header: FC = () => {
  const scrollRef = useRef<null | HTMLDivElement>(null);
  const isMounted = useRef<boolean>(false);
  const isMountedSecond = useRef<boolean>(false);
  const location = useLocation();

  const { itemsCount, totalPrice, items } = useAppSelector(
    selectCart,
    shallowEqual,
  );
  const isLight = useAppSelector(selectIsLight);
  const dispatch = useAppDispatch();

  // scroll effect
  useEffect(() => {
    if (isMountedSecond.current) {
      scrollRef.current?.scrollIntoView();
    }

    isMountedSecond.current = true;
  }, [location.pathname]);

  // local storage logic
  useEffect(() => {
    if (isMounted.current) {
      const jsonCart = JSON.stringify(items);
      const jsonTheme = JSON.stringify(isLight);

      localStorage.setItem('cart', jsonCart);
      localStorage.setItem('isLight', jsonTheme);
    }

    isMounted.current = true;
  }, [items, isLight]);

  const changeThemes = () => {
    dispatch(setTheme(!isLight));
  };

  const keyChangeThemes = (e: KeyboardEvent<SVGElement>) => {
    if (e.code === 'Enter') dispatch(setTheme(!isLight));
  };

  return (
    <HeaderWrapper ref={scrollRef} isCart={location.pathname === '/cart'}>
      <LogoSection />

      <Right>
        <Icon
          as={isLight ? BsFillMoonFill : BsFillSunFill}
          onClick={changeThemes}
          onKeyDown={(e) => keyChangeThemes(e)}
          mode={location.pathname === '/cart' ? 'cart' : ''}
          tabIndex={0}
          data-testid="themeIcon"
          data-theme={isLight ? 'light' : 'dark'}
        />
        {location.pathname !== '/cart' && (
          <ButtonCart to="/cart" as={Link} data-testid="cartLink">
            <span>{totalPrice} $</span>
            <Delimiter />
            <BsCart3 />
            <span>{itemsCount}</span>
          </ButtonCart>
        )}
      </Right>
    </HeaderWrapper>
  );
};
