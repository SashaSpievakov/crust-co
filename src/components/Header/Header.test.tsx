import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Header from './Header';
import rendererWithProviders from '../../tests/helpers/rendererWithProviders';
import renderWithProviders from '../../tests/helpers/renderWithProviders';

describe('Header Test', () => {
  test('renders the Header component', () => {
    const snapshot = rendererWithProviders(<Header />);
    expect(snapshot).toMatchSnapshot();
  });

  test('renders the Header component on the Cart page', () => {
    const snapshot = rendererWithProviders(<Header />, '/cart');
    expect(snapshot).toMatchSnapshot();
  });

  test('checks link redirect to the Cart page', () => {
    renderWithProviders(<Header />, true, '/testing');
    screen.debug();
    const link = screen.getByTestId('cartLink');

    userEvent.click(link);

    expect(
      screen.getByRole('heading', {
        name: /your cart is empty/i,
      }),
    ).toBeInTheDocument();
  });
});
