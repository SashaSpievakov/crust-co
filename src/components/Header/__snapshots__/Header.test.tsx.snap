// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Tests renders the Header component 1`] = `
Object {
  "children": Array [
    <section
      className="sc-dkrFOg sc-eDvSVe gYraVE iRTugX"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <section
          className="sc-bcXHqe bwjlDT"
        >
          <img
            alt="Pizza logo"
            height="54.5"
            src="main-logo.svg"
            width="38"
          />
          <div>
            <h1
              className="sc-gswNZR gWFkxJ"
            >
              Pizza Place
            </h1>
            <p>
              the most delicious pizzas in town
            </p>
          </div>
        </section>
      </a>
      <section
        className="sc-jSUZER jAeLAN"
      >
        <svg
          className="sc-gKPRtg cMcVRX"
          fill="currentColor"
          height="1em"
          mode=""
          onClick={[Function]}
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"
          />
        </svg>
        <a
          className="sc-ftTHYK sc-jrcTuL KXpKQ cKUqLm"
          data-testid="cartLink"
          href="/cart"
          onClick={[Function]}
        >
          <span>
            0
             $
          </span>
          <div
            className="sc-iBYQkv iTPFCD"
          />
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 1.5A.5.5 0 0 1 .5 1H2a.5.5 0 0 1 .485.379L2.89 3H14.5a.5.5 0 0 1 .49.598l-1 5a.5.5 0 0 1-.465.401l-9.397.472L4.415 11H13a.5.5 0 0 1 0 1H4a.5.5 0 0 1-.491-.408L2.01 3.607 1.61 2H.5a.5.5 0 0 1-.5-.5zM3.102 4l.84 4.479 9.144-.459L13.89 4H3.102zM5 12a2 2 0 1 0 0 4 2 2 0 0 0 0-4zm7 0a2 2 0 1 0 0 4 2 2 0 0 0 0-4zm-7 1a1 1 0 1 1 0 2 1 1 0 0 1 0-2zm7 0a1 1 0 1 1 0 2 1 1 0 0 1 0-2z"
            />
          </svg>
          <span>
            0
          </span>
        </a>
      </section>
    </section>,
  ],
  "props": Object {
    "className": "sc-hLBbgP iodrKd",
  },
  "type": "header",
}
`;

exports[`Header Tests renders the Header component on the Cart page 1`] = `
Object {
  "children": Array [
    <section
      className="sc-dkrFOg sc-eDvSVe gYraVE cvniFy"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <section
          className="sc-bcXHqe bwjlDT"
        >
          <img
            alt="Pizza logo"
            height="54.5"
            src="main-logo.svg"
            width="38"
          />
          <div>
            <h1
              className="sc-gswNZR gWFkxJ"
            >
              Pizza Place
            </h1>
            <p>
              the most delicious pizzas in town
            </p>
          </div>
        </section>
      </a>
      <section
        className="sc-jSUZER jAeLAN"
      >
        <svg
          className="sc-gKPRtg jvGLTl"
          fill="currentColor"
          height="1em"
          mode="cart"
          onClick={[Function]}
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"
          />
        </svg>
      </section>
    </section>,
  ],
  "props": Object {
    "className": "sc-hLBbgP iodrKd",
  },
  "type": "header",
}
`;
